File: SpearoGo/Services/TideStore.swift
import Foundation
import CoreLocation

/// Persistent store for tide predictions.
///
/// The app's CacheService is deliberately memory-only — "watch storage is
/// precious" — but a week of tides has to survive a relaunch, or every cold
/// start costs a metered lookup. The backend returns seven days for the price
/// of one, so keeping them is most of the point.
///
/// Small on disk: at most 7 days across a handful of saved spots, and anything
/// whose day has passed is dropped on the next write.
struct TideStore {
    private let defaults = UserDefaults.standard
    private static let prefix = "tides."
    /// Predictions for a given day do not change.
    private static let cacheDuration: TimeInterval = 24 * 3600

    private struct Entry: Codable {
        let savedAt: Date
        var days: [TideData] = []
        /// The backend said this coordinate has no sea. A settled fact.
        var noCoverage: Bool = false
    }

    private func key(_ coordinate: CLLocationCoordinate2D) -> String {
        String(format: "%@%.2f,%.2f", Self.prefix, coordinate.latitude, coordinate.longitude)
    }

    private func read(_ coordinate: CLLocationCoordinate2D) -> Entry? {
        guard let data = defaults.data(forKey: key(coordinate)) else { return nil }
        return try? JSONDecoder().decode(Entry.self, from: data)
    }

    private func write(_ entry: Entry, for coordinate: CLLocationCoordinate2D) {
        guard let data = try? JSONEncoder().encode(entry) else { return }
        defaults.set(data, forKey: key(coordinate))
    }

    /// A cached day, still within its life.
    func fresh(coordinate: CLLocationCoordinate2D, date: String) -> TideData? {
        guard let entry = read(coordinate), !entry.noCoverage,
              Date().timeIntervalSince(entry.savedAt) < Self.cacheDuration else { return nil }
        return rollover(entry.days, on: date)
    }

    /// A cached day past its life, for when the network is gone. Better than
    /// nothing, and flagged in the UI — but never invented.
    func stale(coordinate: CLLocationCoordinate2D, date: String) -> TideData? {
        guard let entry = read(coordinate), !entry.noCoverage else { return nil }
        return rollover(entry.days, on: date)
    }

    /// Whether this coordinate is already known to have no sea.
    func knownWithoutCoverage(coordinate: CLLocationCoordinate2D) -> Bool {
        guard let entry = read(coordinate) else { return false }
        return entry.noCoverage && Date().timeIntervalSince(entry.savedAt) < Self.cacheDuration
    }

    func save(coordinate: CLLocationCoordinate2D, days: [TideData]) {
        let today = TideService.dayKey(for: Date(), utcOffsetSeconds: 0)
        let kept = days.filter { $0.date >= today }
        write(Entry(savedAt: Date(), days: kept), for: coordinate)
    }

    /// Remembered so a landlocked spot stops asking. Held for the same day as
    /// real data — long enough to stop the traffic, short enough that a backend
    /// gaining coverage is picked up.
    func rememberNoCoverage(coordinate: CLLocationCoordinate2D) {
        write(Entry(savedAt: Date(), days: [], noCoverage: true), for: coordinate)
    }
}

/// A day, carrying the following day's turns alongside its own.
///
/// Without this the page reads "—" for the next high from the day's last high
/// until midnight — exactly when tomorrow's dive gets planned. It lives beside
/// the store because both the network and the cache path come through it;
/// applying it in only one of them is how it was missed on Wear.
func rollover(_ days: [TideData], on date: String) -> TideData? {
    guard let index = days.firstIndex(where: { $0.date == date }) else { return nil }
    var day = days[index]
    let next = days.indices.contains(index + 1) ? days[index + 1].events : []
    day.events = day.events + next
    return day
}

File: wear/app/src/main/java/com/spearotracker/spearogo/services/TideStore.kt
package com.spearotracker.spearogo.services

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.spearotracker.spearogo.models.TideData
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Persistent store for tide predictions.
 *
 * The app's CacheService is deliberately memory-only — "watch storage is
 * precious" — but a week of tides has to survive a relaunch, or every cold
 * start costs a metered lookup. The backend returns seven days for the price of
 * one, so keeping them is most of the point.
 *
 * Small on disk: at most 7 days across a handful of saved spots, and anything
 * whose day has passed is dropped on the next write.
 */
@Singleton
class TideStore @Inject constructor(
    @ApplicationContext context: Context
) {
    private val prefs = context.getSharedPreferences("tides", Context.MODE_PRIVATE)
    private val gson = Gson()

    private data class Entry(
        val savedAt: Long,
        val days: List<TideData> = emptyList(),
        /** The backend said this coordinate has no sea. A settled fact. */
        val noCoverage: Boolean = false
    )

    private fun key(latitude: Double, longitude: Double) =
        "%.2f,%.2f".format(latitude, longitude)

    private fun read(latitude: Double, longitude: Double): Entry? {
        val raw = prefs.getString(key(latitude, longitude), null) ?: return null
        return try {
            gson.fromJson(raw, object : TypeToken<Entry>() {}.type)
        } catch (e: Exception) {
            null
        }
    }

    private fun isFresh(entry: Entry) =
        System.currentTimeMillis() - entry.savedAt < CACHE_MILLIS

    /** A cached day, still within its life. */
    fun fresh(latitude: Double, longitude: Double, date: String): TideData? {
        val entry = read(latitude, longitude) ?: return null
        if (entry.noCoverage || !isFresh(entry)) return null
        return entry.days.firstOrNull { it.date == date }
    }

    /**
     * A cached day past its life, for when the network is gone. Better than
     * nothing, and flagged in the UI — but never invented.
     */
    fun stale(latitude: Double, longitude: Double, date: String): TideData? {
        val entry = read(latitude, longitude) ?: return null
        if (entry.noCoverage) return null
        return entry.days.firstOrNull { it.date == date }
    }

    /** Whether this coordinate is already known to have no sea. */
    fun knownWithoutCoverage(latitude: Double, longitude: Double): Boolean {
        val entry = read(latitude, longitude) ?: return false
        return entry.noCoverage && isFresh(entry)
    }

    fun save(latitude: Double, longitude: Double, days: List<TideData>) {
        val today = java.time.LocalDate.now().toString()
        val kept = days.filter { it.date >= today }
        prefs.edit()
            .putString(key(latitude, longitude), gson.toJson(Entry(System.currentTimeMillis(), kept)))
            .apply()
    }

    /**
     * Remembered so a landlocked spot stops asking. Held for the same day as
     * real data — long enough to stop the traffic, short enough that a backend
     * gaining coverage is picked up.
     */
    fun rememberNoCoverage(latitude: Double, longitude: Double) {
        prefs.edit()
            .putString(
                key(latitude, longitude),
                gson.toJson(Entry(System.currentTimeMillis(), emptyList(), noCoverage = true))
            )
            .apply()
    }

    private companion object {
        /** Predictions for a given day do not change. */
        const val CACHE_MILLIS = 24 * 60 * 60 * 1000L
    }
}
